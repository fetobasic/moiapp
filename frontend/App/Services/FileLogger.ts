import { NavigationState } from '@react-navigation/native';
import AppConfig, { env } from 'App/Config/AppConfig';
import { store } from 'App/Store';
import { format } from 'date-fns';
import { Dirs, FileSystem } from 'react-native-file-access';

const MAX_FILE_SIZE = 10 * 1024 * 1024; // 10MB

function FileLogger() {
  let logs: string[] = [];
  let writeToFileInterval: NodeJS.Timeout;
  const FILES_DIR = `${Dirs.DocumentDir}/GZ_Logs`;

  // constructor
  (() => {
    writeToFileInterval = setInterval(() => {
      writeLogsToFile();
    }, AppConfig.writeToFileInterval);
  })();

  // Private methods
  const createFileWithContent = (content: string) =>
    new Promise<void>((resolve, reject) => {
      const fileName = `${format(new Date(), 'yyyyMMdd_HH:mm:ss')}.${env}.log`;

      FileSystem.writeFile(`${FILES_DIR}/${fileName}`, content)
        .then(() => resolve())
        .catch((error) => reject(error));
    });

  const writeLogsToFile = async () => {
    try {
      if (!logs.length) {
        return;
      }

      if (!(await FileSystem.exists(FILES_DIR))) {
        await FileSystem.mkdir(FILES_DIR);
      }

      const files = await FileSystem.ls(FILES_DIR);

      const dataToWrite = logs.join('');
      logs = [];

      if (files.length) {
        const lastFile = files[files.length - 1];

        const stats = await FileSystem.stat(`${FILES_DIR}/${lastFile}`);

        if (stats.size > MAX_FILE_SIZE) {
          await createFileWithContent(dataToWrite);
        } else {
          await FileSystem.appendFile(`${FILES_DIR}/${lastFile}`, dataToWrite);
        }
      } else {
        await createFileWithContent(dataToWrite);
      }
      console?.tron?.log?.('WriteLogsToFile Success');
    } catch (error) {
      console?.tron?.log?.('WriteLogsToFile Error', error);
    }
  };

  // Public methods
  const clearTimer = () => {
    if (writeToFileInterval) {
      clearInterval(writeToFileInterval);
    }
  };

  const addLog = (title: string, message: string) => {
    // Check if Logger is enabled
    const isFileLoggerEnabled = store?.getState().application?.isFileLoggerEnabled;

    if (!isFileLoggerEnabled) {
      return;
    }

    const log = `[${format(new Date(), 'yyyy-MM-dd_HH:mm:ss')}] ::${title}:: ${message}\n`;
    logs.push(log);
  };

  const navigationHandler = (state: NavigationState | undefined) => {
    let message = '';

    if (!state) {
      return;
    }

    const lastHistory = state.history?.[state.history.length - 1] as { type: string; status: string } | undefined;

    if (lastHistory?.type === 'drawer') {
      message += `Drawer: ${lastHistory.status}. `;
    }

    const currentRoute =
      state.routes?.[state.index]?.state?.routes?.[(state.routes?.[state.index]?.state?.routes?.length || 1) - 1];

    if (currentRoute) {
      message += `Current route: ${currentRoute.name}. ${
        currentRoute.params ? `Params: ${JSON.stringify(currentRoute.params)}.` : ''
      }`;
    }

    addLog('NAVIGATION', message);
  };

  return {
    FILES_DIR,
    addLog,
    clearTimer,
    navigationHandler,
  };
}

export const fileLogger = FileLogger();
