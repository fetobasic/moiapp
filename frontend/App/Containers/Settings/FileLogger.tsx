import React, { useState } from 'react';
import { View, Text, FlatList, StyleSheet } from 'react-native';
import { FileSystem } from 'react-native-file-access';
import Share from 'react-native-share';
import { ApplicationStyles, Colors, Fonts, Metrics } from 'App/Themes';
import { HeaderSimple as Header, Pressable } from 'App/Components';
import { useMount } from 'App/Hooks';
import { fileLogger } from 'App/Services/FileLogger';
import { showConfirm } from 'App/Services/Alert';

import WarningIcon from 'App/Images/Icons/warn.svg';

type Props = {
  defaultFiles?: string[];
};

const FileLogger = (props: Props) => {
  const [files, setFiles] = useState<string[]>(props.defaultFiles || []);
  const [selectedFile, setSelectedFile] = useState<string | null>(null);

  useMount(() => {
    FileSystem.ls(fileLogger.FILES_DIR).then(setFiles);
  });

  const deleteFile = async () => {
    if (!selectedFile) {
      return;
    }

    await FileSystem.unlink(`${fileLogger.FILES_DIR}/${selectedFile}`);
    setFiles(files.filter((file) => file !== selectedFile));
    setSelectedFile(null);
  };

  const deleteAllFiles = async () => {
    files.forEach(async (file) => {
      await FileSystem.unlink(`${fileLogger.FILES_DIR}/${file}`);
    });
    setFiles([]);
    setSelectedFile(null);
  };

  const shareFile = async () => {
    if (!selectedFile) {
      return;
    }

    Share.open({
      filename: selectedFile,
      type: 'text/plain',
      url: `file://${fileLogger.FILES_DIR}/${selectedFile}`,
    });
  };

  const renderItem = ({ item }: { item: string }) => {
    return (
      <Pressable testID={`selectFile_${item}`} style={styles.rowSection} onPress={() => setSelectedFile(item)}>
        <View style={[styles.check, selectedFile === item && styles.checkSelected]}>
          {selectedFile === item && <View style={styles.round} />}
        </View>
        <Text style={styles.textFileName}>{item}</Text>
      </Pressable>
    );
  };

  const renderEmpty = () => (
    <View style={styles.sectionEmty}>
      <Text style={Fonts.font.condensed.h3}>No files found</Text>
    </View>
  );

  return (
    <View style={ApplicationStyles.mainContainer}>
      <Header title="Logger Files" />
      <FlatList style={styles.container} data={files} renderItem={renderItem} ListEmptyComponent={renderEmpty} />
      <View style={styles.bottomSection}>
        <Pressable
          testID="deleteFile"
          disabled={!selectedFile}
          style={styles.button}
          onPress={() =>
            showConfirm(
              'Delete',
              `Are you sure you want to delete ${selectedFile}?`,
              deleteFile,
              undefined,
              undefined,
              false,
              <WarningIcon />,
            )
          }>
          <Text style={[styles.textFileName, !selectedFile && styles.buttonDisabled]}>Delete</Text>
        </Pressable>
        <Pressable
          testID="deleteAllFiles"
          disabled={!files.length}
          style={[styles.button, !files.length && styles.buttonDisabled]}
          onPress={() =>
            showConfirm(
              'Delete All',
              'Are you sure you want to delete all files',
              deleteAllFiles,
              undefined,
              undefined,
              false,
              <WarningIcon />,
            )
          }>
          <Text style={styles.textFileName}>Delete All</Text>
        </Pressable>
        <Pressable testID="shareBtn" disabled={!selectedFile} style={styles.button} onPress={() => shareFile()}>
          <Text style={[styles.textFileName, !selectedFile && styles.buttonDisabled]}>Share</Text>
        </Pressable>
      </View>
    </View>
  );
};

const styles = StyleSheet.create({
  container: {
    paddingHorizontal: Metrics.baseMargin,
    paddingVertical: Metrics.marginSection,
  },
  sectionEmty: {
    flex: 1,
    justifyContent: 'center',
    alignItems: 'center',
    marginTop: Metrics.bigMargin * 2,
  },
  rowSection: {
    flexDirection: 'row',
    alignItems: 'center',
    paddingVertical: Metrics.marginSection,
    borderBottomColor: Colors.border,
    borderBottomWidth: 1,
  },
  check: {
    width: 20,
    height: 20,
    borderRadius: 10,
    borderWidth: 2,
    justifyContent: 'center',
    alignItems: 'center',
    borderColor: Colors.transparentWhite('0.87'),
    marginRight: Metrics.baseMargin,
  },
  checkSelected: {
    borderColor: Colors.green,
  },
  round: {
    width: 10,
    height: 10,
    borderRadius: 5,
    backgroundColor: Colors.green,
  },
  textFileName: {
    ...Fonts.style.subtitleOne,
  },
  bottomSection: {
    flexDirection: 'row',
    marginBottom: Metrics.marginSection,
    borderTopColor: Colors.border,
    borderTopWidth: 1,
  },
  button: {
    flex: 1,
    justifyContent: 'center',
    alignItems: 'center',
    padding: Metrics.marginSection,
  },
  buttonDisabled: {
    opacity: 0.5,
  },
});

export default FileLogger;
