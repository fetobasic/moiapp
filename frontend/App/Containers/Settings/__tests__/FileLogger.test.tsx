import React from 'react';
import { render, fireEvent } from '@testing-library/react-native';
import FileLogger from '../FileLogger';

// ls Files Mock: ['file1', 'file2']

describe('FileLogger', () => {
  test('renders screen', () => {
    const tree = render(<FileLogger />).toJSON();
    expect(tree).toMatchSnapshot();
  });
  test('delete file', () => {
    const { getByTestId } = render(<FileLogger defaultFiles={['file1', 'file2']} />);

    const selectFile = getByTestId('selectFile_file1');
    fireEvent.press(selectFile);

    const deleteFile = getByTestId('deleteFile');
    fireEvent.press(deleteFile);
  });

  test('delete all files', () => {
    const { getByTestId } = render(<FileLogger defaultFiles={['file1', 'file2']} />);

    const deleteFile = getByTestId('deleteAllFiles');
    fireEvent.press(deleteFile);
  });

  test('share File', () => {
    const { getByTestId } = render(<FileLogger defaultFiles={['file1', 'file2']} />);

    const selectFile = getByTestId('selectFile_file1');
    fireEvent.press(selectFile);

    const shareBtn = getByTestId('shareBtn');
    fireEvent.press(shareBtn);
  });

  test('share non selected File', () => {
    const { getByTestId } = render(<FileLogger />);

    const shareBtn = getByTestId('shareBtn');
    fireEvent.press(shareBtn);
  });
});
