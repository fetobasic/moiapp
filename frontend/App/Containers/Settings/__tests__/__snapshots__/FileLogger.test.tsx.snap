// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FileLogger renders screen 1`] = `
<View
  style={
    {
      "backgroundColor": "rgba(27, 27, 27, 1)",
      "flex": 1,
    }
  }
>
  <View
    style={
      [
        {
          "backgroundColor": "rgba(27, 27, 27, 1)",
          "borderBottomLeftRadius": 24,
          "borderBottomRightRadius": 24,
          "elevation": 12,
          "shadowColor": "rgba(0, 0, 0, 1)",
          "shadowOffset": {
            "height": 4,
            "width": 4,
          },
          "shadowOpacity": 0.2,
          "shadowRadius": 12,
          "zIndex": 1,
        },
        {
          "paddingTop": 0,
        },
      ]
    }
  >
    <View
      style={
        {
          "alignItems": "center",
          "flexDirection": "row",
          "height": 60,
          "justifyContent": "center",
          "marginHorizontal": 24,
        }
      }
    >
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        hitSlop={20}
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          [
            {
              "opacity": 1,
            },
            {
              "left": 0,
              "padding": 10,
              "position": "absolute",
            },
          ]
        }
      >
        <SvgMock />
      </View>
      <View
        style={
          {
            "alignItems": "center",
            "marginHorizontal": 48,
          }
        }
      >
        <Text
          numberOfLines={1}
          style={
            {
              "color": "rgba(255, 255, 255, 1)",
              "fontFamily": "RobotoCondensed-Medium",
              "fontSize": 20,
              "fontWeight": "500",
              "lineHeight": 32,
            }
          }
        >
          Logger Files
        </Text>
      </View>
    </View>
  </View>
  <RCTScrollView
    ListEmptyComponent={[Function]}
    data={[]}
    getItem={[Function]}
    getItemCount={[Function]}
    keyExtractor={[Function]}
    onContentSizeChange={[Function]}
    onLayout={[Function]}
    onMomentumScrollBegin={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={0.0001}
    stickyHeaderIndices={[]}
    style={
      {
        "paddingHorizontal": 24,
        "paddingVertical": 16,
      }
    }
    viewabilityConfigCallbackPairs={[]}
  >
    <View>
      <View
        style={
          {
            "alignItems": "center",
            "flex": 1,
            "justifyContent": "center",
            "marginTop": 64,
          }
        }
      >
        <Text
          style={
            {
              "color": "rgba(255, 255, 255, 1)",
              "fontFamily": "RobotoCondensed-Medium",
              "fontSize": 20,
              "fontWeight": "500",
              "lineHeight": 32,
            }
          }
        >
          No files found
        </Text>
      </View>
    </View>
  </RCTScrollView>
  <View
    style={
      {
        "borderTopColor": "rgba(54, 54, 54, 1)",
        "borderTopWidth": 1,
        "flexDirection": "row",
        "marginBottom": 16,
      }
    }
  >
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": true,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        [
          {
            "opacity": 1,
          },
          {
            "alignItems": "center",
            "flex": 1,
            "justifyContent": "center",
            "padding": 16,
          },
        ]
      }
      testID="deleteFile"
    >
      <Text
        style={
          [
            {
              "color": "rgba(255, 255, 255, 1)",
              "fontFamily": "Roboto-Medium",
              "fontSize": 16,
              "fontWeight": "400",
              "lineHeight": 24,
            },
            {
              "opacity": 0.5,
            },
          ]
        }
      >
        Delete
      </Text>
    </View>
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": true,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        [
          {
            "opacity": 1,
          },
          [
            {
              "alignItems": "center",
              "flex": 1,
              "justifyContent": "center",
              "padding": 16,
            },
            {
              "opacity": 0.5,
            },
          ],
        ]
      }
      testID="deleteAllFiles"
    >
      <Text
        style={
          {
            "color": "rgba(255, 255, 255, 1)",
            "fontFamily": "Roboto-Medium",
            "fontSize": 16,
            "fontWeight": "400",
            "lineHeight": 24,
          }
        }
      >
        Delete All
      </Text>
    </View>
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": true,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        [
          {
            "opacity": 1,
          },
          {
            "alignItems": "center",
            "flex": 1,
            "justifyContent": "center",
            "padding": 16,
          },
        ]
      }
      testID="shareBtn"
    >
      <Text
        style={
          [
            {
              "color": "rgba(255, 255, 255, 1)",
              "fontFamily": "Roboto-Medium",
              "fontSize": 16,
              "fontWeight": "400",
              "lineHeight": 24,
            },
            {
              "opacity": 0.5,
            },
          ]
        }
      >
        Share
      </Text>
    </View>
  </View>
</View>
`;
